🐛 Destination AWS Datalake: fix KeyError issue and add parquet support

henriblancke opened the pull request:
## What
*Describe what the change is solving*
* Resolve destination-aws-datalake bug and add more functionality for destination
  * Resolve error when a stream schema does not have properties defined for `"type":"object"`, resulting in a  `KeyError('properties')` and breaking the sync
  * Add support for parquet format, file compression and partitioning

## How
*Describe the solution*

This change uses `awswrangler` to get a lot of additional functionality for this destination out of the box, such as:
* Support different output formats (json, parquet)
* Support output file compression
   * Support gzip, snappy and zstd compression

Additional features include:
* Support partitioning based on cursor fields
   * Partition on date, year, month, day or year/month/day
* Improve pyarrow/athena type casting:
   * Use the source `json-schema` to infer pyarrow and athena types
   * Support complex data types such as structs and arrays and apply sane defaults when they can't be inferred.
* Allow using `namespaces` so a single destination can create multiple databases (useful for example, when you want multiple sources to use their own database).
* Configure the use of `governed` vs `external` tables
* Configure how floats should be handled in athena (`double` vs `decimal`).
* Allow assigning default LF-tags to databases created by this connection (helps with permission management)
* Delete table in athena when a connection is reset.
* Remove the test database after validating connection and permissions in `check`.

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

There are some spec and configuration changes

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>
<img width="1609" alt="Screen Shot 2022-09-26 at 8 19 58 PM" src="https://user-images.githubusercontent.com/1708162/192403259-3ae71c55-f855-426b-b477-82922668f224.png">
</details>

<details><summary><strong>Integration</strong></summary>
<img width="1601" alt="Screen Shot 2022-09-26 at 8 01 06 PM" src="https://user-images.githubusercontent.com/1708162/192403171-712702bd-52e2-45a2-9bb4-260378f6dfc1.png">
</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/airbytehq/airbyte?pullRequest=17193) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/airbytehq/airbyte?pullRequest=17193) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/airbytehq/airbyte?pullRequest=17193) it.</sub>

marcosmarxm commented:
Sorry the delay in review this amazing contribution @henriblancke !!! Hope to get a review until wednesday.

marcosmarxm commented:
@henriblancke sorry the delay here. I didn't have time this week and I'm OOO Today. I added this to my priority to-do list on Monday.

henriblancke commented:
@marcosmarxm thanks for the update and thanks for adding this to your priority list for Monday 😄! Enjoy your day off today!

marcosmarxm reviewed (commented):
@henriblancke tests are failing to me, looks I can't find the database for integration tests. I need some test to check what is causing it.

henriblancke commented:
@marcosmarxm sorry to hear that! What test is failing for you? Do you have `secrets/config.json` set up with a real AWS account that has permissions to lake formation, s3 and athena? 

All integration tests seem to pass for me locally. I may be able to help troubleshoot, let me know what test(s) are failing for you. Thanks!

henriblancke commented:
@marcosmarxm any updates? Let me know if I can help.

henriblancke commented:
@marcosmarxm I made a few changes in the past few weeks, but things should be in their final state now.

marcosmarxm commented:
Hello :wave:, first thank you for this amazing contribution. 

We really appreciate the effort you've made to improve the project.
We ask you patience for the code review. Last month our team was focused on Hacktoberfest event and that probably left some PR without the proper feedback. And this week, due to the Thanksgiving US Holiday, most our team is out of office with their families. Another important piece of information why code won't be merge this week is: as a safety measure the core team has decided to freeze merging code to main branch to keep the release stable. Next week we'll return to you with the proper code review and update the status of your contribution.

If you have any questions feel free to send me a message in Slack!
Thanks!

marcosmarxm commented:
@henriblancke Sorry about the lack of response here. I'll start the process of reviewing it! Thanks for the contribution

marcosmarxm commented:
/test connector=connectors/destination-aws-datalake

marcosmarxm commented:
/test connector=connectors/destination-aws-datalake

marcosmarxm commented:
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

sh4sh commented:
/test connector=connectors/destination-aws-datalake

sh4sh commented:
Unit tests passed on my local machine! The CI creds will need to be modified to support the new config.json.

I updated the config.json to test locally, however `check` still fails on my local machine because of an `AlreadyExistsException` on `CreateTable`. Also, `test_write` is failing on `delete_table_if_exists`. I've attached the integration test output below.

@henriblancke were you able to get the integration tests to pass when the table already exists in the destination?

It's possible our CI creds don't have the right access, although presumably they worked before...? I'll look tomorrow in case something's going on there.

```
(.venv) sunny@YXCQY6DK7W destination-aws-datalake % python -m pytest integration_tests
================================================ FAILURES ================================================
________________________________________ test_check_valid_config _________________________________________

config = <redacted>

    def test_check_valid_config(config: Mapping):
        outcome = DestinationAwsDatalake().check(logger, config)
>       assert outcome.status == Status.SUCCEEDED
E       AssertionError: assert <Status.FAILED: 'FAILED'> == <Status.SUCCEEDED: 'SUCCEEDED'>
E         +<Status.FAILED: 'FAILED'>
E         -<Status.SUCCEEDED: 'SUCCEEDED'>

integration_tests/integration_test.py:79: AssertionError
------------------------------------------- Captured log call --------------------------------------------
ERROR    awswrangler._utils:_utils.py:353 Retrying <function _overwrite_table at 0x11ede7550> | Fail number 1/3 | Exception: An error occurred (AlreadyExistsException) when calling the CreateTable operation: Table already exists.
ERROR    awswrangler._utils:_utils.py:353 Retrying <function _overwrite_table at 0x11ede7550> | Fail number 2/3 | Exception: An error occurred (AlreadyExistsException) when calling the CreateTable operation: Table already exists.
ERROR    airbyte:destination.py:151 Could not create a table in database airbyte-integration: AlreadyExistsException('An error occurred (AlreadyExistsException) when calling the CreateTable operation: Table already exists.')
_______________________________________________ test_write _______________________________________________

config = <redacted>
configured_catalog = ConfiguredAirbyteCatalog(streams=[ConfiguredAirbyteStream(stream=AirbyteStream(name='append_stream', json_schema={'typ...cremental'>, cursor_field=None, destination_sync_mode=<DestinationSyncMode.overwrite: 'overwrite'>, primary_key=None)])

    def test_write(config: Mapping, configured_catalog: ConfiguredAirbyteCatalog):
        """
        This test verifies that:
            1. writing a stream in "overwrite" mode overwrites any existing data for that stream
            2. writing a stream in "append" mode appends new records without deleting the old ones
            3. The correct state message is output by the connector at the end of the sync
        """
        append_stream, overwrite_stream = configured_catalog.streams[0].stream.name, configured_catalog.streams[1].stream.name

        # make sure we start with empty tables
        for tbl in [append_stream, overwrite_stream]:
>           wr.catalog.delete_table_if_exists(database=config["lakeformation_database_name"], table=tbl)

integration_tests/integration_test.py:114:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.9/site-packages/awswrangler/_config.py:472: in wrapper
    return function(**args)
.venv/lib/python3.9/site-packages/awswrangler/catalog/_delete.py:85: in delete_table_if_exists
    client_glue: boto3.client = _utils.client(service_name="glue", session=boto3_session)
.venv/lib/python3.9/site-packages/awswrangler/_config.py:472: in wrapper
    return function(**args)
.venv/lib/python3.9/site-packages/awswrangler/_utils.py:116: in client
    return ensure_session(session=session).client(
.venv/lib/python3.9/site-packages/boto3/session.py:299: in client
    return self._session.create_client(
.venv/lib/python3.9/site-packages/botocore/session.py:976: in create_client
    client = client_creator.create_client(
.venv/lib/python3.9/site-packages/botocore/client.py:155: in create_client
    client_args = self._get_client_args(
.venv/lib/python3.9/site-packages/botocore/client.py:485: in _get_client_args
    return args_creator.get_client_args(
.venv/lib/python3.9/site-packages/botocore/args.py:92: in get_client_args
    final_args = self.compute_client_args(
.venv/lib/python3.9/site-packages/botocore/args.py:205: in compute_client_args
    endpoint_config = self._compute_endpoint_config(
.venv/lib/python3.9/site-packages/botocore/args.py:313: in _compute_endpoint_config
    return self._resolve_endpoint(**resolve_endpoint_kwargs)
.venv/lib/python3.9/site-packages/botocore/args.py:418: in _resolve_endpoint
    return endpoint_bridge.resolve(
.venv/lib/python3.9/site-packages/botocore/client.py:590: in resolve
    resolved = self.endpoint_resolver.construct_endpoint(
.venv/lib/python3.9/site-packages/botocore/regions.py:229: in construct_endpoint
    result = self._endpoint_for_partition(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <botocore.regions.EndpointResolver object at 0x11ed7aaf0>
partition = OrderedDict([('defaults', OrderedDict([('hostname', '{service}.{region}.{dnsSuffix}'), ('protocols', ['https']), ('sig...ict([('variants', [OrderedDict([('hostname', 'xray-fips.us-west-2.amazonaws.com'), ('tags', ['fips'])])])]))]))]))]))])
service_name = 'glue', region_name = None, use_dualstack_endpoint = None, use_fips_endpoint = None
force_partition = False

    def _endpoint_for_partition(
        self,
        partition,
        service_name,
        region_name,
        use_dualstack_endpoint,
        use_fips_endpoint,
        force_partition=False,
    ):
        partition_name = partition["partition"]
        if (
            use_dualstack_endpoint
            and partition_name in self._UNSUPPORTED_DUALSTACK_PARTITIONS
        ):
            error_msg = (
                "Dualstack endpoints are currently not supported"
                " for %s partition" % partition_name
            )
            raise EndpointVariantError(tags=['dualstack'], error_msg=error_msg)

        # Get the service from the partition, or an empty template.
        service_data = partition['services'].get(
            service_name, DEFAULT_SERVICE_DATA
        )
        # Use the partition endpoint if no region is supplied.
        if region_name is None:
            if 'partitionEndpoint' in service_data:
                region_name = service_data['partitionEndpoint']
            else:
>               raise NoRegionError()
E               botocore.exceptions.NoRegionError: You must specify a region.

.venv/lib/python3.9/site-packages/botocore/regions.py:277: NoRegionError
======================================== short test summary info =========================================
FAILED integration_tests/integration_test.py::test_check_valid_config - AssertionError: assert <Status....
FAILED integration_tests/integration_test.py::test_write - botocore.exceptions.NoRegionError: You must ...
====================================== 2 failed, 2 passed in 34.99s ======================================
```

henriblancke commented:
@sh4sh thanks! I'll take a look at that today

henriblancke commented:
@sh4sh this is most likely related to access and configuration. The check creates and deletes a table and it looks like it fails here because no default aws region is defined? The configuration has changed slightly vs the previous version of this destination.

henriblancke commented:
@sh4sh this is [an example config](https://github.com/henriblancke/airbyte/blob/destination-aws-datalake-parquet/airbyte-integrations/connectors/destination-aws-datalake/unit_tests/fixtures/config.json). Let me know if there is anything I can help with?

sh4sh commented:
Hey @henriblancke , yeah I did use the above format for the tests with the addition of `bucket_prefix`. Can you make sure this is supported? Here is the config I'm using:

```
{
    "aws_account_id": "<redacted>",
    "credentials": {
        "credentials_title": "<redacted>",
        "aws_access_key_id": "<redacted>",
        "aws_secret_access_key": "<redacted>"
    },
    "region": "us-east-1",
    "bucket_name": "airbyte-datalake-useast1",
    "bucket_prefix": "test",
    "lakeformation_database_name": "airbyte-integration",
    "format": {
      "format_type": "Parquet",
      "compression_codec": "SNAPPY"
    },
    "partitioning": "NO PARTITIONING"
}
```

I tried removing `region` to see what would happen, we get a different error on `test_check_valid_config`:

```
================================================================== short test summary info ===================================================================
FAILED integration_tests/integration_test.py::test_check_valid_config - botocore.exceptions.NoRegionError: You must specify a region.
FAILED integration_tests/integration_test.py::test_write - botocore.exceptions.NoRegionError: You must specify a region.
================================================================ 2 failed, 2 passed in 25.21s ================================================================
```
Here's the error with a configured region:

```
============================================================================================ short test summary info =============================================================================================
FAILED integration_tests/integration_test.py::test_check_valid_config - AssertionError: assert <Status.FAILED: 'FAILED'> == <Status.SUCCEEDED: 'SUCCEEDED'>
FAILED integration_tests/integration_test.py::test_write - botocore.exceptions.NoRegionError: You must specify a region.
========================================================================================== 2 failed, 2 passed in 49.13s ==========================================================================================
```

Additionally, here are the results of running `check` with region configured:

```
(.venv) sunny@YXCQY6DK7W destination-aws-datalake % python main.py check --config secrets/config.json
{"type": "LOG", "log": {"level": "ERROR", "message": "Retrying <function _overwrite_table at 0x137bfe8b0> | Fail number 1/3 | Exception: An error occurred (AlreadyExistsException) when calling the CreateTable operation: Table already exists."}}
{"type": "LOG", "log": {"level": "ERROR", "message": "Retrying <function _overwrite_table at 0x137bfe8b0> | Fail number 2/3 | Exception: An error occurred (AlreadyExistsException) when calling the CreateTable operation: Table already exists."}}
{"type": "LOG", "log": {"level": "ERROR", "message": "Could not create a table in database airbyte-integration: AlreadyExistsException('An error occurred (AlreadyExistsException) when calling the CreateTable operation: Table already exists.')"}}
{"type": "CONNECTION_STATUS", "connectionStatus": {"status": "FAILED", "message": "Could not create a table in database airbyte-integration: AlreadyExistsException('An error occurred (AlreadyExistsException) when calling the CreateTable operation: Table already exists.')"}}
```

To verify that the credentials are configured properly, I ran `check` on the current version of destination-aws-datalake and it was successful:

```
`(.venv) sunny@YXCQY6DK7W destination-aws-datalake % python main.py check --config secrets/config.json
{"type": "LOG", "log": {"level": "DEBUG", "message": "Starting Lakeformation Transaction"}}
{"type": "LOG", "log": {"level": "DEBUG", "message": "Transaction id = 110b95cf0b7f42f5b559b1ad0cbd1ad4"}}
{"type": "LOG", "log": {"level": "DEBUG", "message": "Exiting LakeformationTransaction context manager"}}
{"type": "LOG", "log": {"level": "DEBUG", "message": "{\"ResponseMetadata\": {\"RequestId\": \"42667e23-bbae-4767-afd9-b520ab5f8e07\", \"HTTPStatusCode\": 200, \"HTTPHeaders\": {\"date\": \"Thu, 09 Feb 2023 19:32:14 GMT\", \"content-type\": \"application/json\", \"content-length\": \"173\", \"connection\": \"keep-alive\", \"x-amzn-requestid\": \"42667e23-bbae-4767-afd9-b520ab5f8e07\"}, \"RetryAttempts\": 0}, \"TransactionDescription\": {\"TransactionId\": \"110b95cf0b7f42f5b559b1ad0cbd1ad4\", \"TransactionStatus\": \"active\", \"TransactionStartTime\": \"2023-02-09 14:32:12.905000-05:00\", \"TransactionEndTime\": \"1969-12-31 19:00:00-05:00\"}}"}}
{"type": "LOG", "log": {"level": "DEBUG", "message": "Exiting LakeformationTransaction context manager due to reaching end of with block"}}
{"type": "LOG", "log": {"level": "DEBUG", "message": "Commiting Lakeformation Transaction 110b95cf0b7f42f5b559b1ad0cbd1ad4"}}
{"type": "CONNECTION_STATUS", "connectionStatus": {"status": "SUCCEEDED"}}`
```

henriblancke commented:
@sh4sh I found the issue, the integration tests were using some direct `aws-wrangler` calls that weren't use the boto session created by the connector. It automatically used my local aws creds that's why the tests passed for me locally. In the process I also discovered [this issue](https://github.com/aws/aws-sdk-pandas/issues/2007). Can you give running the tests and `check` another try? Thank you!

henriblancke commented:
@sh4sh let me know if there's anything else I can do to help here

sh4sh commented:
The NoRegionError is gone! 😁 But still seeing some failures, can you try running `python -m pytest integration_tests` locally and check the results?

Results here:
<details>
<summary>Test results</summary>

```
(.venv) sunny@YXCQY6DK7W destination-aws-datalake % python -m pytest integration_tests
==================================================================== FAILURES =====================================================================
_____________________________________________________________ test_check_valid_config _____________________________________________________________

config = {'aws_account_id': '<redacted>', 'bucket_name': '<redacted>', 'bucket_prefix': '<redacted>', 'credentials': '<redacted>', 'aws_secret_access_key': '<redacted>', 'credentials_title': '<redacted>'}, ...}

    def test_check_valid_config(config: Mapping):
        outcome = DestinationAwsDatalake().check(logger, config)
>       assert outcome.status == Status.SUCCEEDED
E       AssertionError: assert <Status.FAILED: 'FAILED'> == <Status.SUCCEEDED: 'SUCCEEDED'>
E         +<Status.FAILED: 'FAILED'>
E         -<Status.SUCCEEDED: 'SUCCEEDED'>

integration_tests/integration_test.py:81: AssertionError
---------------------------------------------------------------- Captured log call ----------------------------------------------------------------
ERROR    awswrangler._utils:_utils.py:353 Retrying <function _overwrite_table at 0x1186c6550> | Fail number 1/3 | Exception: An error occurred (AlreadyExistsException) when calling the CreateTable operation: Table already exists.
ERROR    awswrangler._utils:_utils.py:353 Retrying <function _overwrite_table at 0x1186c6550> | Fail number 2/3 | Exception: An error occurred (AlreadyExistsException) when calling the CreateTable operation: Table already exists.
ERROR    airbyte:destination.py:147 Could not create table airbyte_test in database airbyte-integration: AlreadyExistsException('An error occurred (AlreadyExistsException) when calling the CreateTable operation: Table already exists.')
___________________________________________________________________ test_write ____________________________________________________________________

config = {'aws_account_id': '<redacted>', 'bucket_name': '<redacted>', 'bucket_prefix': '<redacted>', 'credentials': '<redacted>', 'aws_secret_access_key': '<redacted>', 'credentials_title': '<redacted>'}, ...}
configured_catalog = ConfiguredAirbyteCatalog(streams=[ConfiguredAirbyteStream(stream=AirbyteStream(name='append_stream', json_schema={'typ...cremental'>, cursor_field=None, destination_sync_mode=<DestinationSyncMode.overwrite: 'overwrite'>, primary_key=None)])

    def test_write(config: Mapping, configured_catalog: ConfiguredAirbyteCatalog):
        """
        This test verifies that:
            1. writing a stream in "overwrite" mode overwrites any existing data for that stream
            2. writing a stream in "append" mode appends new records without deleting the old ones
            3. The correct state message is output by the connector at the end of the sync
        """
        append_stream, overwrite_stream = configured_catalog.streams[0].stream.name, configured_catalog.streams[1].stream.name

        destination = DestinationAwsDatalake()

        connector_config = ConnectorConfig(**config)
        aws_handler = AwsHandler(connector_config, destination)

        database = connector_config.lakeformation_database_name

        # make sure we start with empty tables
        for tbl in [append_stream, overwrite_stream]:
            aws_handler.reset_table(database, tbl)

        first_state_message = _state({"state": "1"})

        first_record_chunk = [_record(append_stream, str(i), i, datetime.now()) for i in range(5)] + [
            _record(overwrite_stream, str(i), i, datetime.now()) for i in range(5)
        ]

        second_state_message = _state({"state": "2"})
        second_record_chunk = [_record(append_stream, str(i), i, datetime.now()) for i in range(5, 10)] + [
            _record(overwrite_stream, str(i), i, datetime.now()) for i in range(5, 10)
        ]

        expected_states = [first_state_message, second_state_message]
>       output_states = list(
            destination.write(
                config, configured_catalog, [*first_record_chunk, first_state_message, *second_record_chunk, second_state_message]
            )
        )

integration_tests/integration_test.py:137:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <destination_aws_datalake.destination.DestinationAwsDatalake object at 0x128cd21c0>
config = {'aws_account_id': '<redacted>', 'bucket_name': '<redacted>', 'bucket_prefix': '<redacted>', 'credentials': '<redacted>', 'aws_secret_access_key': '<redacted>', 'credentials_title': '<redacted>'}, ...}
configured_catalog = ConfiguredAirbyteCatalog(streams=[ConfiguredAirbyteStream(stream=AirbyteStream(name='append_stream', json_schema={'typ...cremental'>, cursor_field=None, destination_sync_mode=<DestinationSyncMode.overwrite: 'overwrite'>, primary_key=None)])
input_messages = [AirbyteMessage(type=<Type.RECORD: 'RECORD'>, log=None, spec=None, connectionStatus=None, catalog=None, record=Airbyte...tetime.datetime(2023, 2, 21, 18, 36, 22, 371516), 'string_col': None}, emitted_at=0, namespace=None), state=None), ...]

    def write(
        self, config: Mapping[str, Any], configured_catalog: ConfiguredAirbyteCatalog, input_messages: Iterable[AirbyteMessage]
    ) -> Iterable[AirbyteMessage]:

        """
        Reads the input stream of messages, config, and catalog to write data to the destination.

        This method returns an iterable (typically a generator of AirbyteMessages via yield) containing state messages received
        in the input message stream. Outputting a state message means that every AirbyteRecordMessage which came before it has been
        successfully persisted to the destination. This is used to ensure fault tolerance in the case that a sync fails before fully completing,
        then the source is given the last state message output from this method as the starting point of the next sync.

        :param config: dict of JSON configuration matching the configuration declared in spec.json
        :param configured_catalog: The Configured Catalog describing the schema of the data being received and how it should be persisted in the
                                    destination
        :param input_messages: The stream of input messages received from the source
        :return: Iterable of AirbyteStateMessages wrapped in AirbyteMessage structs
        """
        connector_config = ConnectorConfig(**config)

        try:
            aws_handler = AwsHandler(connector_config, self)
        except ClientError as e:
            logger.error(f"Could not create session due to exception {repr(e)}")
            raise Exception(f"Could not create session due to exception {repr(e)}")

        # creating stream writers
        streams = {
            s.stream.name: StreamWriter(aws_handler=aws_handler, config=connector_config, configured_stream=s)
            for s in configured_catalog.streams
        }

        for message in input_messages:
            if message.type == Type.STATE:
                if not message.state.data:

                    if message.state.stream:
                        stream = message.state.stream.stream_descriptor.name
                        logger.info(f"Received empty state for stream {stream}, resetting stream")
                        if stream in streams:
                            streams[stream].reset()
                        else:
                            logger.warning(f"Trying to reset stream {stream} that is not in the configured catalog")

                    if not message.state.stream:
                        logger.info("Received empty state for, resetting all streams including non-incremental streams")
                        for stream in streams:
                            streams[stream].reset()

                # Flush records when state is received
>               if message.state.stream:
E               AttributeError: 'AirbyteStateMessage' object has no attribute 'stream'

destination_aws_datalake/destination.py:79: AttributeError
---------------------------------------------------------------- Captured log call ----------------------------------------------------------------
INFO     airbyte:aws.py:217 Resetting table airbyte-integration.append_stream
INFO     airbyte:aws.py:217 Resetting table airbyte-integration.overwrite_stream
INFO     airbyte:stream_writer.py:33 Creating StreamWriter for airbyte-integration:append_stream
INFO     airbyte:stream_writer.py:33 Creating StreamWriter for airbyte-integration:overwrite_stream
================================================================ warnings summary =================================================================
.venv/lib/python3.9/site-packages/jsonschema/compat.py:6
.venv/lib/python3.9/site-packages/jsonschema/compat.py:6
  /Users/sunny/git/pullrequests/17193/airbyte/airbyte-integrations/connectors/destination-aws-datalake/.venv/lib/python3.9/site-packages/jsonschema/compat.py:6: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
    from collections import MutableMapping, Sequence  # noqa

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================= short test summary info =============================================================
FAILED integration_tests/integration_test.py::test_check_valid_config - AssertionError: assert <Status.FAILED: 'FAILED'> == <Status.SUCCEEDED: '...
FAILED integration_tests/integration_test.py::test_write - AttributeError: 'AirbyteStateMessage' object has no attribute 'stream'
==================================================== 2 failed, 2 passed, 2 warnings in 35.53s =====================================================
```

</details>

Also, is check passing on your local machine? I am still getting a similar error:

```
(.venv) sunny@YXCQY6DK7W destination-aws-datalake % python main.py check --config secrets/config.json
{"type": "LOG", "log": {"level": "ERROR", "message": "Could not create table airbyte_test in database airbyte-integration: AlreadyExistsException('An error occurred (AlreadyExistsException) when calling the CreateTable operation: Table already exists.')"}}
{"type": "CONNECTION_STATUS", "connectionStatus": {"status": "FAILED", "message": "Could not create table airbyte_test in database airbyte-integration: AlreadyExistsException('An error occurred (AlreadyExistsException) when calling the CreateTable operation: Table already exists.')"}}
```

henriblancke commented:
@sh4sh I see, the connector currently assumes that the table doesn't exist when you first run it. The first time it runs a check, it'll create a table and then remove that table and it's data using `aws_handler.reset_table`. I can change the logic and run `aws_handler.reset_table` before it tries to create and write to the table?

henriblancke commented:
@sh4sh should be fixed now

henriblancke commented:
@sh4sh `check` now works for me even when the table exists. Airbyte does need permissions to delete it from the glue catalog. I'll rerun the integration tests here locally and keep you posted.

sh4sh commented:
@henriblancke in [my previous comment](https://github.com/airbytehq/airbyte/pull/17193#issuecomment-1424715023) I verified that `check` was passing with previous connector versions, however I'll still look into it. Do you know if any additional permissions would be required for this iteration?

henriblancke commented:
@sh4sh  The integration tests pass for me, but this issue in your output `FAILED integration_tests/integration_test.py::test_write - AttributeError: 'AirbyteStateMessage' object has no attribute 'stream'` makes me think we're on different cdk versions potentially so let me verify that.

```
(destination-aws-datalake) ➜  destination-aws-datalake git:(destination-aws-datalake-parquet) ✗ python -m pytest integration_tests
======================================================================================================================== test session starts =========================================================================================================================
platform darwin -- Python 3.9.13, pytest-6.2.5, py-1.11.0, pluggy-1.0.0 -- /Users/henriblancke/opt/anaconda3/envs/destination-aws-datalake/bin/python
cachedir: .pytest_cache
rootdir: /Users/henriblancke/Documents/personal/airbyte, configfile: pytest.ini
collected 4 items                                                                                                                                                                                                                                                    

integration_tests/integration_test.py::test_check_valid_config {"type": "LOG", "log": {"level": "INFO", "message": "Resetting table test_db.airbyte_test"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Deleting table test_db.airbyte_test"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Deleting objects in s3://wisetack-feature26-us-east-1-data-lake/test/test_db/airbyte_test/"}}
PASSED
integration_tests/integration_test.py::test_check_invalid_aws_region_config {"type": "LOG", "log": {"level": "ERROR", "message": "not_a_real_region is not a valid AWS region"}}
PASSED
integration_tests/integration_test.py::test_check_invalid_aws_account_config PASSED
integration_tests/integration_test.py::test_write {"type": "LOG", "log": {"level": "INFO", "message": "Resetting table test_db.append_stream"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Resetting table test_db.overwrite_stream"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Creating StreamWriter for test_db:append_stream"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Creating StreamWriter for test_db:overwrite_stream"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Resetting table test_db.append_stream"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Deleting table test_db.append_stream"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Deleting objects in s3://wisetack-feature26-us-east-1-data-lake/test/test_db/append_stream/"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Resetting table test_db.overwrite_stream"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Deleting table test_db.overwrite_stream"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Deleting objects in s3://wisetack-feature26-us-east-1-data-lake/test/test_db/overwrite_stream/"}}
PASSED

========================================================================================================================= 4 passed in 41.62s =========================================================================================================================
```

The previous version of the connector seems to [create the table only once](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/destination-aws-datalake/destination_aws_datalake/aws.py#L75-L103) and doesn't ever get rid of it and/or validate that it can actually write to it. Can you verify your aws user has the `glue:DeleteTable` and `glue:DeleteDatabase` permissions?

sh4sh commented:
I see, thanks for the info @henriblancke. I'll check our creds and update when I know more.

...considering the last iteration didn't do a thorough `check`, just want to send you some extra big thanks for this contribution 😅

henriblancke commented:
@sh4sh on `AttributeError: 'AirbyteStateMessage' object has no attribute 'stream'` shouldn't `AirbyteStateMessage` always have an attribute `stream` that should default to `None`?

Ooh absolutely no worries, it's been a lot of fun. Sorry for all the back and forth :)

grishick commented:
draft PR to run CI: https://github.com/airbytehq/airbyte/pull/23855

grishick reviewed (changes requested):
The integration tests are not passing. Here is my latest run https://github.com/airbytehq/airbyte/actions/runs/4359907106/jobs/7622249697

henriblancke commented:
@grishick it looks like `format` is not defined in your config, should be flagged as required in `spec.json` now

henriblancke commented:
@grishick for the connector icon check, there was no icon defined in the previous version. Should I go ahead and define one?

grishick commented:
> @grishick for the connector icon check, there was no icon defined in the previous version. Should I go ahead and define one?

yes please

grishick commented on airbyte-integrations/connectors/destination-aws-datalake/destination_aws_datalake/spec.json:
```diff
@@ -6,20 +6,22 @@
+      "region",
+      "bucket_name",
+      "lakeformation_database_name",
+      "format"
```
adding new required spec fields will make this a breaking change - any connection that already runs with an older version will break upon upgrade. Is it possible to adde defaults for `format`, `aws_account_id`,  and `lakeformation_database_name` to avoid making this a breaking change?

henriblancke replied:
I'll add defaults for `format` and `lakeformation_database_name`

henriblancke replied:
actually, it looks like `lakeformation_database_name` didn't have a default in the previous version and wasn't flagged as required. It most likely should be required, the connector needs to know about what database it should use by default?

grishick commented on airbyte-integrations/connectors/destination-aws-datalake/destination_aws_datalake/spec.json:
```diff
@@ -6,20 +6,22 @@
     "type": "object",
-    "required": ["credentials", "region", "bucket_name", "bucket_prefix"],
+    "required": [
+      "aws_account_id",
```
given that we already ask for `credentials`, why do we need a separate `aws_account_id` ?

henriblancke replied:
yep that's a good point, it was defined so I left it there but I can remove it from required.

henriblancke commented:
@grishick let me know what the next steps here are. Thanks again for all the help!

grishick commented on airbyte-integrations/connectors/destination-aws-datalake/destination_aws_datalake/config_reader.py:
```diff
@@ -5,10 +5,82 @@
+        elif s == "YEAR/MONTH/DAY":
+            return PartitionOptions.YEAR_MONTH_DAY
+        else:
+            raise ValueError(f"Unknown partition option: {s}")
```
since this is not a required config attribute, it would be better to fall back to the default option `PartitionOptions.NONE` in the `else` part in case `s` is empty. Right now, this change will make any existing connections fail.

grishick commented on airbyte-integrations/connectors/destination-aws-datalake/destination_aws_datalake/config_reader.py:
```diff
@@ -5,10 +5,82 @@
+        elif s == "JSONL":
+            return OutputFormat.JSONL
+        else:
+            raise ValueError(f"Unknown output format: {s}")
```
for backwards compatibility, let's fall back to one of the values in case `s` is empty.

grishick commented on airbyte-integrations/connectors/destination-aws-datalake/destination_aws_datalake/config_reader.py:
```diff
@@ -5,10 +5,82 @@
+        elif str == "UNCOMPRESSED":
+            return CompressionCodec.UNCOMPRESSED
+        else:
+            raise ValueError(f"Unknown compression codec: {str}")
```
similar ask here, let's fall back to `UNCOMPRESSED` in case `str` is empty

grishick commented on airbyte-integrations/connectors/destination-aws-datalake/destination_aws_datalake/destination.py:
```diff
@@ -89,23 +123,28 @@ def check(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> AirbyteConn
-            )
-        if table is None:
-            message = f"Could not create a table in database {connector_config.lakeformation_database_name}"
+        tbl = "airbyte_test"
```
this table name should be randomized and `check` should also attempt to delete this table before exiting. Right now, subsequent test runs fail with this error:
```
	 ERROR    airbyte:destination.py:147 Could not create table airbyte_*** in database airbyte-integration: AlreadyExistsException('An error occurred (AlreadyExistsException) when calling the CreateTable operation: Table already exists.')
	 =========================== short *** summary info ============================
	 FAILED integration_***s/integration_***.py::***_check_valid_config - Asser...
	 ========================= 1 failed, 3 passed in 48.70s =========================

```

henriblancke replied:
Hmm weird, `aws_handler.reset_table(db, tbl)` a couple of lines below this one should delete the table if it exists and all it's underlying data in s3 before it tries to create it again. Does your configured iam role or user have permissions to access/delete the table? The integration test logs show the following sequence of events:
1. A call made to GetTable returning "Table not found" (user may not have permissions to access it)
2. A call made to CreateTable returning "Table already exists" (error could be caused by table existing but user not being able to access it)
3. A call made to DeleteTable returning "Table not found"
Makes me think that for some reason the role or user does not have permissions to access the existing table? But I may be overlooking something? I'll make sure to randomize the table name in my next commit

grishick replied:
A scenario that the integration test should account for is that a previous run did not properly cleanup (it could be cancelled, interrupted, failed due to a hosting problem, CI problem or anything else).

Another problem is that `CHECK` should not assume that `airbyte_test` table does not exist in the destination, because a user could be using multiple connections to write to the same destination. It is safer to add a random suffix/preffix to the tables created by `CHECK` method to avoid a name collision.

henriblancke replied:
That makes sense!

grishick commented:
> @grishick let me know what the next steps here are. Thanks again for all the help!

@henriblancke I re-ran integration tests and added several more comments based on the failures. I modified the config that we have saved in GSM in order to get passed failures caused by missing config values, and added one more comment about a failure in `check` method.

henriblancke commented:
> > @grishick let me know what the next steps here are. Thanks again for all the help!
> 
> @henriblancke I re-ran integration tests and added several more comments based on the failures. I modified the config that we have saved in GSM in order to get passed failures caused by missing config values, and added one more comment about a failure in `check` method.

@grishick thanks again for the review 🚀, I really appreciate all the help here! I responded to your `check` comment, added a random element to the table name and addressed the other comments.

grishick commented:
Rerunning the integration tests: https://github.com/airbytehq/airbyte/pull/23855

grishick commented:
/test connector=connectors/destination-aws-datalake

grishick commented:
Published and merged here: https://github.com/airbytehq/airbyte/pull/23855

henriblancke commented:
> Published and merged here: #23855

Thanks again for all the help @grishick!
